### PR TITLE
Add dir and compressed cache types

### PR DESCRIPTION
**Pull request by martindurant**
Dir: Particularly for nested zarr and parquet directory structure
Compressed: for single and multiple files, each of which may be expandable.

### CODE REVIEW

File: intake/source/cache.py [hunk 1]
```diff
@@ -238,6 +241,55 @@ def _download(file_in, file_out, blocksize, output=False):
...
+                files_out2.append(fout)
+        return files_in2, files_out2
+
+
+class CompressedCache(BaseCache):
+    def _make_files(self, urlpath, **kwargs):
+        import dask.bytes.compression
+        import tempfile
+        ext = os.path.splitext(urlpath)[1]
+        comp = kwargs.get('compression', ext)
```
Comment (mmccarty) [inline intake/source/cache.py]: Maybe inherit from `FileCache` so you can unpack then call its `_make_files`?

Comment (martindurant) [reply intake/source/cache.py]: I haven't figured out what to do yet. The dir-cache is maybe ready, but needs tests, of course.

File: intake/source/cache.py [hunk 2]
```diff
@@ -123,18 +123,18 @@ def load(self, urlpath, output=None):
...
         cache_paths = []
-        files_in = open_files(urlpath, 'rb', **self._storage_options)
-        files_out = [open_files([self._path(f.path, subdir)], 'wb')[0]
-                     for f in files_in]
+        files_in, files_out = self._make_files(urlpath)
+        self._load(files_in, files_out, cache_paths, urlpath)
+
+        return cache_paths
+
+    def _load(self, files_in, files_out, cache_paths, urlpath):
```
[the remark this answers is not in the document]
Comment (mmccarty) [reply intake/source/cache.py]: This implementation can be moved to `FileCache` .

File: intake/source/cache.py [hunk 3]
```diff
@@ -34,7 +34,7 @@ def sanitize_path(path):
 display = set()
 
 
-class FileCache(object):
+class BaseCache(object):
```
Comment (mmccarty) [inline intake/source/cache.py]: Add abstract methods for `_load` and `_make_dirs`.

File: intake/source/tests/cached.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        regex: '{{ CATALOG_DIR }}'
+  dirs:
+    driver: textfiles
+    args:
+      urlpath: '{{ CATALOG_DIR }}/main'
+    cache:
+      - type: dir
+        regex: '{{ CATALOG_DIR }}'
+        argkey: urlpath
+        depth: 2
```
Comment (mmccarty) [inline intake/source/tests/cached.yaml]: Maybe test another large-ish depth?

File: intake/source/cache.py [hunk 5]
```diff
@@ -123,18 +123,18 @@ def load(self, urlpath, output=None):
         """
         if conf.get('cache_disabled', False):
             return [urlpath]
-        output = output if output is not None else conf.get(
+        self.output = output if output is not None else conf.get(
```
[the remark this answers is not in the document]
Comment (martindurant) [reply intake/source/cache.py]: Would not have the desired behaviour when `output=False` (as opposed to `None`).

[the remark this answers is not in the document]
Comment (mmccarty) [reply intake/source/cache.py]: Oh, I see.  Yep, never mind.

**Comment by martindurant** [comment]
@mmccarty , anything still to do here?

File: intake/source/cache.py [hunk 6]
```diff
@@ -240,6 +249,119 @@ def _download(file_in, file_out, blocksize, output=False):
         display.remove(out)
 
 
+class FileCache(BaseCache):
+    """Cache specific set of files
+
+    Input is a single file URL, URL with glob characters or list of URLs. Output
+    is a specific set of local files.
+    """
+
```
Comment (mmccarty) [inline intake/source/cache.py]: Should `BaseCache._load` go here too?

Comment (martindurant) [reply intake/source/cache.py]: I don't think so - subclasses may want to call this basic download-a-file, as both Dir and Compressed do.

Comment (mmccarty) [reply intake/source/cache.py]: Okay, I didn't realize that.  thanks

**Comment by mmccarty** [comment]
@martindurant LGTM!

**Pull request merged**